File: 2025/12/03/structured-outputs-llms-reliable-json/python/src/validator.py
"""Validate JSON against Pydantic schemas."""
import json
from typing import Type, TypeVar
from pydantic import BaseModel, ValidationError

T = TypeVar('T', bound=BaseModel)


def validate_json(
    json_data: dict,
    schema: Type[T]
) -> T:
    """
    Validate JSON data against a Pydantic schema.
    
    Args:
        json_data: JSON data to validate
        schema: Pydantic model class
    
    Returns:
        Validated model instance
    
    Raises:
        ValueError: If validation fails
    """
    try:
        return schema.model_validate(json_data)
    except ValidationError as e:
        # Format errors for better readability
        errors = []
        for error in e.errors():
            field_path = ".".join(str(loc) for loc in error["loc"])
            errors.append({
                "field": field_path,
                "message": error["msg"],
                "type": error["type"],
                "input": error.get("input")
            })
        
        error_msg = f"Validation failed: {json.dumps(errors, indent=2)}"
        raise ValueError(error_msg) from e


def format_validation_error(error: ValidationError) -> str:
    """
    Format a ValidationError into a readable string.
    
    Args:
        error: ValidationError instance
    
    Returns:
        Formatted error message
    """
    errors = []
    for err in error.errors():
        field_path = ".".join(str(loc) for loc in err["loc"])
        errors.append(f"{field_path}: {err['msg']}")
    
    return "; ".join(errors)

File: 2025/12/03/structured-outputs-llms-reliable-json/typescript/src/validator.ts
import { z } from 'zod';
import { ZodSchema } from 'zod';

/**
 * Validate JSON data against a Zod schema.
 */
export function validateJson<T>(
  jsonData: unknown,
  schema: ZodSchema<T>
): T {
  const result = schema.safeParse(jsonData);
  
  if (!result.success) {
    const errors = result.error.errors.map(err => ({
      field: err.path.join('.'),
      message: err.message,
      code: err.code,
    }));
    
    throw new Error(`Validation failed: ${JSON.stringify(errors, null, 2)}`);
  }
  
  return result.data;
}

/**
 * Format a Zod error into a readable string.
 */
export function formatValidationError(error: z.ZodError): string {
  return error.errors
    .map(err => `${err.path.join('.')}: ${err.message}`)
    .join('; ');
}

